pkg/apis/intevents/v1alpha1/brokercell_defaults.go

```diff
@@ -35,22 +63,81 @@ func (bc *BrokerCell) SetDefaults(ctx context.Context) {
+}
+
+func (componentParams *ComponentParameters) SetResourceDefaults(defaultCPURequest, defaultCPULimit, defaultMemoryRequest string, memoryLimitToRequestCoefficient float64) {
+	componentParams.SetCPUDefaults(defaultCPURequest, defaultCPULimit)
+	componentParams.SetMemoryDefaults(memoryLimitToRequestCoefficient, defaultMemoryRequest)
+}
+
+// SetCPUDefaults sets the CPU consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetCPUDefaults(defaultCPURequest, defaultCPULimit string) {
+	isRequestSpecified := componentParams.Resources.Requests.CPU != nil
+	isLimitSpecified := componentParams.Resources.Limits.CPU != nil
+	if !isRequestSpecified && !isLimitSpecified {
+		componentParams.Resources.Requests.CPU = ptr.String(defaultCPURequest)
+		componentParams.Resources.Limits.CPU = ptr.String(defaultCPULimit)
+	} else if !isRequestSpecified {
+		componentParams.Resources.Requests.CPU = ptr.String(*componentParams.Resources.Limits.CPU)
 	}
-	if bcs.Components.Fanout.MaxReplicas == nil {
-		bcs.Components.Fanout.MaxReplicas = ptr.Int32(maxReplicas)
+}
+
+// SetMemoryDefaults sets the memory consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetMemoryDefaults(memoryLimitToRequestCoefficient float64, defaultMemoryRequest string) {
+	isRequestSpecified := componentParams.Resources.Requests.Memory != nil
+	isLimitSpecified := componentParams.Resources.Limits.Memory != nil
+	autoSelectLimit := false
+	if !isRequestSpecified && !isLimitSpecified {
+		componentParams.Resources.Requests.Memory = ptr.String(defaultMemoryRequest)
+		autoSelectLimit = true
+	} else if !isRequestSpecified {
+		componentParams.Resources.Requests.Memory = ptr.String(*componentParams.Resources.Limits.Memory)
+	} else if !isLimitSpecified {
+		autoSelectLimit = true
+	}
+	if autoSelectLimit {
+		requestedMemoryQuantity, parsingError := resource.ParseQuantity(*componentParams.Resources.Requests.Memory)
+		if parsingError == nil {
+			autoSelectedLimit := resourceutil.MultiplyQuantity(requestedMemoryQuantity, memoryLimitToRequestCoefficient)
+			componentParams.Resources.Limits.Memory = ptr.String(autoSelectedLimit.String())
+		}
 	}
-	if bcs.Components.Retry.MinReplicas == nil {
-		bcs.Components.Retry.MinReplicas = ptr.Int32(minReplicas)
+}
+
+// SetAutoScalingDefaults sets the autoscaling-related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetAutoScalingDefaults(targetMemoryUsageCoefficient float64, avgCPUUtilization int32) {
+	if componentParams.MinReplicas == nil {
+		componentParams.MinReplicas = ptr.Int32(minReplicas)
 	}
-	if bcs.Components.Retry.MaxReplicas == nil {
-		bcs.Components.Retry.MaxReplicas = ptr.Int32(maxReplicas)
+	if componentParams.MaxReplicas == nil {
+		componentParams.MaxReplicas = ptr.Int32(maxReplicas)
 	}
```

Comment:
I know this is existing logic. But it has the same problem - if I (as a user) only set the min, it's possible my min is greater than the default max, which is wrong. As I mentioned in the other comment, if we have a HPA spec, then we can easily tell if a value is explicitly left blank by the user or not.

Reply:
I've noticed that and thought it might be best not to change the existing structure -- to keep changes focused (already quite many) and, AFAIK, there is an ongoing work that is dependent on current format in perf/test area. But I have added a validation to account for possible min-max inconsistency.